packages/core/src/lib/engine.core.ts

```diff
@@ -262,38 +262,113 @@ export class DiagramEngineCore {
-        const zoomFactor = xFactor < yFactor ? xFactor : yFactor;
-
-        let newZoomLvl =
-          diagramModel.getZoomLevel() * (zoomFactor - additionalZoomFactor);
-        const maxZoomOut = diagramModel.getMaxZoomOut();
-
-        if (maxZoomOut && newZoomLvl < maxZoomOut) {
-          newZoomLvl = maxZoomOut;
-        }
+  getNodeLayersRect(nodes: NodeModel[], margin = 0) {
+    if (!nodes || nodes.length === 0) {
+      return { width: 0, height: 0, top: 0, left: 0 };
+    }
 
-        diagramModel.setZoomLevel(newZoomLvl);
+    const firstNode = nodes[0];
+    const { x: firstNodeXCoords, y: firstNodeYCoords } = firstNode.getCoords();
 
-        // TODO: either block the canvas movement on 0,0 or detect the top left furthest element and set the offest to its edges
-        diagramModel.setOffset(0, 0);
-      });
+    let rightNodeWidth = firstNode.getWidth();
+    let bottomNodeHeight = firstNode.getHeight();
+
+    let minX = firstNodeXCoords;
+    let maxX = firstNodeXCoords;
+    let minY = firstNodeYCoords;
+    let maxY = firstNodeYCoords;
+
+    // Go over all nodes and calc the min, max points of all of them
+    nodes.forEach(node => {
+      const { x, y } = node.getCoords();
+      if (x < minX) minX = x;
+      if (x > maxX) {
+        maxX = x;
+        rightNodeWidth = node.getWidth();
+      }
+      if (y < minY) minY = y;
+      if (y > maxY) {
+        maxY = y;
+        bottomNodeHeight = node.getHeight();
+      }
+    })
+
+    // Calc the rect, adding the desired margin.
+    return {
+      width: (maxX - minX + rightNodeWidth) + (2 * margin),
+      height: (maxY - minY + bottomNodeHeight) + (2 * margin),
+      top: minY - margin,
+      left: minX - margin
+    }
+  }
+
+  /**
+   * Set zoom and offset so that all nodes of the diagram will be seen at view
+   */
+  zoomToFit(margin = 100) {
+    const diagramModel = this.getDiagramModel();
+    if (diagramModel) {
+      const allNodes = diagramModel.getNodesArray();
+      if (allNodes?.length > 0) {
+        this.zoomToNodes(allNodes, margin);
+      }
+    }
+  }
+
+  /**
+  * fit the canvas zoom levels and offset to the nodes contained.
+  * @param nodes set zoom and offset so that all those node will be seen at view
+  * @param margin allow for further zooming out to make sure edges doesn't cut (in px)
+  */
+  zoomToNodes(nodes: NodeModel[], margin = 100) {
+    if (nodes?.length > 0) {
+      const diagramModel = this.getDiagramModel();
+      if (diagramModel) {
+        this.canvas$.value$
+          .pipe(
+            filter(
+              (canvas: HTMLElement | null | undefined): canvas is HTMLElement =>
+                canvas !== null && canvas !== undefined
+            ),
+            take(1),
+            delay(0)
+          )
+          .subscribe((canvas) => {
+            // Get nodes layers bounding rect with the desired margin
+            const nodesLayersRect = this.getNodeLayersRect(nodes, margin);
```

Comment:
and if you throw in the above function then throwing here could be correct as well, or returning with a console.warn